### ISSUE TITLE
Issue when updating from mongoose 4.4.2 to 4.4.3

### ISSUE THREAD

**Issue by benjaminpageaud** [Automattic/mongoose#3874]
Hello guys,

We found a bug when updating mongoose to 4.4.3 (in fact we had a "latest" tag in our package.json, so it has downloaded the 4.4.3) : 
```javascript
/app/node_modules/mongoose/node_modules/mongodb/lib/server.js:283               
      process.nextTick(function() { throw err; })                               
                                    ^                                           

TypeError: Cannot read property 'collectionName' of undefined                   
    at Query.findOne (/app/node_modules/mongoose/node_modules/mquery/lib/mquery.
js:1678:36)                                                                     
    at Config.onPackageForSettingsLoaded (/app/node_modules/meanio/lib/core_modu
les/config/index.js:62:11)                                                      
    at Config.onDatabaseLoadPackageModel (/app/node_modules/meanio/lib/core_modu
les/config/index.js:122:3)                                                      
    at Consumer.Dependable.runAction (/app/node_modules/meanio/node_modules/lazy
-dependable/index.js:73:22)                                                     
    at Consumer.Dependable.fire (/app/node_modules/meanio/node_modules/lazy-depe
ndable/index.js:70:53)                                                          
    at Consumer.onResolved (/app/node_modules/meanio/node_modules/lazy-dependabl
e/index.js:120:8)                                                               
    at Consumer.Dependable.resolve (/app/node_modules/meanio/node_modules/lazy-d
ependable/index.js:56:10)                                                       
    at Meanio.Container.resolve (/app/node_modules/meanio/node_modules/lazy-depe
ndable/index.js:195:5)                                                          
    at Config.loadPackageModel (/app/node_modules/meanio/lib/core_modules/config
/index.js:111:32)                                                               
    at Config.loadSettings (/app/node_modules/meanio/lib/core_modules/config/ind
ex.js:55:22)                                                                    
    at Consumer.Dependable.runAction (/app/node_modules/meanio/node_modules/lazy
-dependable/index.js:73:22)                                                     
    at Consumer.Dependable.fire (/app/node_modules/meanio/node_modules/lazy-depe
ndable/index.js:70:53)                                                          
    at Consumer.onResolved (/app/node_modules/meanio/node_modules/lazy-dependabl
e/index.js:120:8)                                                               
    at Consumer.Dependable.resolve (/app/node_modules/meanio/node_modules/lazy-d
ependable/index.js:56:10)                                                       
    at Meanio.Container.notifyResolved (/app/node_modules/meanio/node_modules/la
zy-dependable/index.js:209:7)                                                   
    at Dependency.onResolved (/app/node_modules/meanio/node_modules/lazy-dependa
ble/index.js:105:18)     
```
After several hours of research, we went back to 4.4.2 and everything worked fine again.
It might be one of your dependencies which causes the problem but we did not dig enough to figure it out.

Our versions : 
NPM : `2.14.12`
Node : `4.3.0`
MongoDB :  `2.6.10`

Have a nice day.

**Comment by vkarpov15** [comment]
Hmm do you have any sample code that can help me repro this? Only thing the stack trace tells me is that you're using meanio and for some reason that mquery function's context doesn't have an `_collection` property

**Comment by benjaminpageaud** [comment]
I'm afraid I have no relevant code to show you.

We deployed three days ago, it worked and, we redeployed today with only small changes in the HTML part (and strictly html, we checked multiple times) and it did not work.

The only thing that "really" changed between the two deployments was mongoose version :(
I would really like to provide you some code but in our situation it would be useless, even if I know it does not help you.

**Comment by vkarpov15** [comment]
Perhaps try opening up an issue on meanio's github? I'll see if I can repro early this week. Can you repro this outside of your prod environment? If so, can you use something like [longjohn](http://npmjs.org/package/longjohn) to get a long stack trace so we can at least see what's triggering all those function calls?

**Comment by normark** [comment]
I have the exact same problem, also related to findOne(...). I had to lock the version at 4.4.2 to avoid issues.

**Comment by normark** [comment]
It is very easy to reproduce on my end. On 4.4.2 the following code runs without errors (except the kerberos warning), however an exception occurs on 4.4.3.

Code to reproduce:
```
var m   = require('mongoose');

var db  = m.connect('mongodb://localhost:27017/mtest');

m.connection.on('connected', function () {
  var UserSchema = new m.Schema({
      someProperty : { type : String, required : true }
  });

  var User = db.model('User', UserSchema);

  User.findOne({ someProperty : 'test' }, function(err, user){
    return;
  });
});
```

Exception thrown on Mongoose 4.4.3:
```
TypeError: Cannot read property 'collectionName' of undefined
    at Query.findOne (/Users/Apps/mongoose-error-repro/node_modules/mongoose/node_modules/mquery/lib/mquery.js:1678:36)
    at NativeConnection.<anonymous> (/Users/Apps/mongoose-error-repro/index.js:13:8)
    at emitNone (events.js:67:13)
    at NativeConnection.emit (events.js:166:7)
    at NativeConnection.Object.defineProperty.set (/Users/Apps/mongoose-error-repro/node_modules/mongoose/lib/connection.js:116:12)
    at open (/Users/Apps/mongoose-error-repro/node_modules/mongoose/lib/connection.js:502:22)
    at NativeConnection.Connection.onOpen (/Users/Apps/mongoose-error-repro/node_modules/mongoose/lib/connection.js:520:5)
    at /Users/Apps/mongoose-error-repro/node_modules/mongoose/lib/connection.js:476:11
    at /Users/Apps/mongoose-error-repro/node_modules/mongoose/lib/drivers/node-mongodb-native/connection.js:60:5
    at /Users/Apps/mongoose-error-repro/node_modules/mongoose/node_modules/mongodb/lib/db.js:227:5
    at connectHandler (/Users/Apps/mongoose-error-repro/node_modules/mongoose/node_modules/mongodb/lib/server.js:280:7)
    at g (events.js:260:16)
    at emitOne (events.js:77:13)
    at emit (events.js:169:7)
    at /Users/Apps/mongoose-error-repro/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/topologies/server.js:504:23
    at commandCallback (/Users/Apps/mongoose-error-repro/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/topologies/server.js:936:9)
/Users/Apps/mongoose-error-repro/node_modules/mongoose/node_modules/mongodb/lib/server.js:283
      process.nextTick(function() { throw err; })
```

**Comment by chrism4111** [comment]
I have the same issue after upgrade to 4.4.3. The following code on 4.4.2 runs very well
My code:

```
const mongoose = require('mongoose'); 
    const databaseSchema = require('./config/database/schema');
    
    let connectionString = 'mongodb://localhost:27017/local';
    
    angular.module('elixir').service('dashboardService', ['$q', DashboardService])
    
    
    function DashboardService($q){
        
        return {
            getSettings: getSettings
        }
        
        function getSettings(){
            var deferred = $q.defer(),
                conn = ClientConnection(),
                settings = conn.model('settings', databaseSchema.getSchema('settings'));
            
            settings.find((err, doc) => {
                deferred.resolve(doc);
            });
            
            return deferred.promise;
        }
    }
    
    function ClientConnection(){
        return mongoose.connect(connectionString);
    }
```

And schema module:

```
const mongoose = require('mongoose');

const Schemas = {
    "settings": {
        "databases": [],
        "selectedDatabase": "String"
    }
}

function getSchema(modelName){
    if(Schemas.hasOwnProperty(modelName)){
        return new mongoose.Schema(Schemas[modelName]);
    }
    
    console.error('Schema for model: ' + modelName + ' not found');
}

module.exports = {
            "getSchema": getSchema
        };
```

On Mongoose 4.4.3 i've got the following exception:

```
TypeError: Cannot read property 'collectionName' of undefined
    at Query.find (C:\Programming\Elixir\elixircrm\app\node_modules\mquery\lib\mquery.js:1623:33)
    at Object.getSettings (file:///C:/Programming/Elixir/elixircrm/app/scripts/dashboard/dashboardService.js:24:22)
    at getSettings (file:///C:/Programming/Elixir/elixircrm/app/scripts/dashboard/dashboardController.js:14:30)
    at new DashboardController (file:///C:/Programming/Elixir/elixircrm/app/scripts/dashboard/dashboardController.js:11:9)
    at Object.instantiate (file:///C:/Programming/Elixir/elixircrm/bower_components/angular/angular.js:4619:14)
    at file:///C:/Programming/Elixir/elixircrm/bower_components/angular/angular.js:9870:28
    at link (file:///C:/Programming/Elixir/elixircrm/bower_components/angular-route/angular-route.js:1001:26)
    at invokeLinkFn (file:///C:/Programming/Elixir/elixircrm/bower_components/angular/angular.js:9492:9)
    at nodeLinkFn (file:///C:/Programming/Elixir/elixircrm/bower_components/angular/angular.js:8978:11)
    at compositeLinkFn (file:///C:/Programming/Elixir/elixircrm/bower_components/angular/angular.js:8226:13)
    at publicLinkFn (file:///C:/Programming/Elixir/elixircrm/bower_components/angular/angular.js:8106:30)
    at file:///C:/Programming/Elixir/elixircrm/bower_components/angular/angular.js:8447:29
    at boundTranscludeFn (file:///C:/Programming/Elixir/elixircrm/bower_components/angular/angular.js:8244:16)
    at controllersBoundTransclude (file:///C:/Programming/Elixir/elixircrm/bower_components/angular/angular.js:9020:20)
    at update (file:///C:/Programming/Elixir/elixircrm/bower_components/angular-route/angular-route.js:959:25)
    at Scope.$broadcast (file:///C:/Programming/Elixir/elixircrm/bower_components/angular/angular.js:17143:28) <ng-view class="ng-scope" data-ng-animate="1">
```

**Comment by vkarpov15** [comment]
@normark thanks for the detailed repro instructions. Looks like even full-time JS devs sometimes forget to curry functions in loops. Fix will be in 4.4.4 tomorrow. In the meantime you can just do:

```javascript
var m   = require('mongoose');

m.connect('mongodb://localhost:27017/mtest');

m.connection.on('connected', function () {
  var UserSchema = new m.Schema({
      someProperty : { type : String, required : true }
  });

  var User = mongoose.model('User', UserSchema);

  User.findOne({ someProperty : 'test' }, function(err, user){
    return;
  });
});
```

Does the same thing, just sidesteps this particular bug.
